data_store/redis.go

```diff
@@ -0,0 +1,64 @@
+package data_store
+
+import (
+	"github.com/go-redis/redis"
+)
+
+const ADDR = "localhost:6379"
+
+type Redis struct {
+	Client *redis.Client
+}
+
+func initRedisClient() *redis.Client {
+	return redis.NewClient(&redis.Options{Addr: ADDR})
+}
+
+func CreateRedis() *Redis {
+	r := Redis{Client: initRedisClient()}
+	if r.Client == nil {
+		return nil
+	}
+
+	return &r
+}
+
+func (r *Redis) AddActivity(username string, req string) {
+	checkExist := r.Client.HExists(username, "path1").Val()
+	if !checkExist {
+		r.createNewClientActivity(username, req)
+	} else {
+		r.updateClientActivity(username, req)
+	}
```

Comment:
What is the advantage of saving the actions of a user in a hash dedicated to each username?

Reply:
Again scalability , say in the future we would collect a lot of data on each user, and want to store and access it efficiently.

Reply:
but you don't need hash for that - you can access each key in Redis in O(1)

Reply:
yes but without hash for each user then we would have independent keys for different user attributes?

**From the redis tutorial:** 
_For example, a User object with name and age fields. Instead of using different keys for name and age, it’s a lot more efficient to use a single hash that contains all the required fields. Instead of initializing different <key>:<id>:<field> string keys such as userhash:1001:name, you can simply create it using {“userhash:1001”: {“name”: “Alice”}}._

_In fact, the official documentation recommends using hashes whenever possible, as small hashes are encoded in a very small space. Hashes are capable of storing many elements and can also be used to store other instances, other than objects._